rpc, server: add DRPC support for node join RPC

cthumuluru-crdb opened the pull request:
DRPC is currently disabled by default and can be enabled via the `rpc.experimental_drpc.enabled` cluster setting. Some code paths, such as command-line tools, don't have access to this setting. This change introduces a `--use-new-drpc` flag for start commands to enable DRPC in those cases. This change enables the use of DRPC for the node join RPC when the `--use-new-rpc` flag is set, and defaults to gRPC otherwise.

Epic: CRDB-55200, CRDB-51459
Fixes: #155592, #149600
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/157993)
<!-- Reviewable:end -->

shubhamdhama commented on pkg/rpc/drpc.go:
```diff
@@ -336,3 +224,173 @@ func newDRPCPeerOptions(
+	rpcCtx *Context,
+) func(ctx context.Context, target string, _ rpcbase.ConnectionClass, additionalOpts ...drpcclient.DialOption) (drpc.Conn, error) {
+	return rpcCtx.drpcDialRaw
+}
```
Is this needed now?

shubhamdhama commented on pkg/rpc/drpc.go:
```diff
@@ -42,104 +40,19 @@ func (d *drpcCloseNotifier) CloseNotify(ctx context.Context) <-chan struct{} {
+	}
+
+	return nil
 }
```
This should go in a separate commit.

cthumuluru-crdb replied:
This is no longer needed after dialer refactoring. Join command dials a one-off connection similar to gRPC.

shubhamdhama reviewed (commented):
LGTM! I wish some non-essential refactor of `drpc.go` could be done in a separate commit even if that's done in this PR. But now I have gone through the PR so it's fine here.

cthumuluru-crdb commented on pkg/rpc/rpcbase/nodedialer.go:
```diff
@@ -16,7 +16,7 @@ import (
 )
 
-var envExperimentalDRPCEnabled = envutil.EnvOrDefaultBool("COCKROACH_EXPERIMENTAL_DRPC_ENABLED", false)
+var envExperimentalDRPCEnabled = envutil.EnvOrDefaultBool("COCKROACH_EXPERIMENTAL_DRPC_ENABLED", true)
```
I usually use `do-not-merge` label for the same reason. I go through the diffs before merging the changes. Keeping a checklist seems a good idea but that also means keeping that list upto date. I'll give it a try.

cthumuluru-crdb committed:
cli,rpc: add --use-new-rpc cli flag to enable DRPC

DRPC is currently disabled by default and can be enabled via the
`rpc.experimental_drpc.enabled` cluster setting. Some code paths, such as
command-line tools, don't have access to this setting. This change introduces
a `--use-new-drpc` flag for start commands to enable DRPC in those cases.

Epic: CRDB-55200
Fixes: #155592
Release note: None

cthumuluru-crdb committed:
rpc: refactor DRPC conn dial to use dialer options

This change replaces hard-coded DRPC conn dial behavior with configurable
dialer options. Dialer options make it easier to customize conn dialing
behavior.

Epic: CRDB-51459
Informs: #149600
Release note: None

cthumuluru-crdb committed:
rpc, server: add DRPC support for node join RPC

This change enables the use of DRPC for the node join RPC when the
`--use-new-rpc` flag is set, and defaults to gRPC otherwise.

Epic: CRDB-51459
Fixed: #149600
Release note: None

shubhamdhama reviewed (approved):
🚀

cthumuluru-crdb commented:
bors r+
